proposals/0015-partitioned-epoch-reward-distribution.md

```diff
@@ -0,0 +1,260 @@
+formula to avoid using floating point value arithmetic:
+
+```
+M = ((4096 - 1)+num_stake_accounts)/4096
+```
+
+### `EpochRewards` Sysvar Account
+
+`EpochRewards` sysvar account records the total rewards plus some other reward
+related information internally. And the account balance reflects the amount of
+pending rewards to distribute.
+
+The layout of `EpochRewards` sysvar is shown in the following pseudo code.
+
+```
+struct RewardRewards{
+   total_reward_in_lamport: u64,          // total rewards for this epoch
+   distributed_reward_in_lamport: u64,    // already distributed reward amount
+   num_stake_accounts: u64,  // number of reward receiving stake accounts
+   num_vote_accounts: u64,   // number of reward receiving vote accounts
+   total_vote_points: u128,  // accrued vote points in the epoch
+}
+
+```
+
+The `EpochRewards` is updated at the start of the first block of the epoch
+(before any transactions are processed), as both the total epoch rewards and
+vote account rewards become available at this time. The
+`distributed_reward_in_lamport` field is updated per reward distribution for
+each block in the reward distribution phase.
+
+### Reward Distribution
+
+Reward distribution phase happens after reward computation phase, which starts
+after the first block in the epoch for this proposal. It lasts for `M` blocks.
+Each of the `M` blocks is responsible for distributing the reward from one
+partition of the rewards from the `EpochRewards` account.
+
+Once all rewards have been distributed, the balance of the `EpochRewards`
+account MUST be reduced to `0` (or something has gone wrong). For safety, Any
+extra lamports in `EpochRewards` accounts will be burned after reward
+distribution phase.
+
+Before each reward distribution, the `EpochRewards` account's `balance` is
+checked to make sure there is enough balance to distribute the reward. After a
+reward is distributed, the balance in `EpochRewards` account is decreased by the
+amount of the reward that was distributed.
+
+### RPC API Changes
+
+To accommodate the new reward distribution approach, there are a few changes
+to the rewards related RPC APIs.
+
+In particular, `getInflationReward` RPC API will change when query the reward
+for the preceding epoch. Such queries will be unavailable during reward period,
+and only available after reward distribution completes.
+
+To help to determine when the reward period completes, a new RPC API,
+`getNumOfRewardDistributionBlocks`,  will be added. This API will returns:
+
+   - `numOfRewardDistributionBlocks: u64`
+
+To query the reward received for individual account, user Apps will first make
+a query to `getNumOfRewardDistributionBlocks`, and wait for that many
+of blocks since the start of the epoch. Then, the App can call
+`getInflationReward` to query the rewards as before.
+
+`getNumOfRewardDistributionBlocks` will also be exposed via sysvar/syscall to
+support on chain Dapp program usage.
```

Comment:
I think we can remove these new RPC methods from the proposal.  

A `getInflationReward` method which returns an error for the previous epoch until the distribution phase completes should be sufficient for most front-ends. The distribution period is expected to be on the order of minutes so they can simply retry in 30 seconds until success. 

If they need more resolution on distribution progress, the `EpochRewards` sysvar can be fetched and its data processed. This is pretty easy to do so by the front-end, so adding more RPC methods currently seems unnecessary for this purpose

Reply:
Removed.

Reply:
Fine to remove as long as the sysvar contains more information -- right now it just as `total_rewards_in_lamport` and its current lamports!